### PR TITLE
Runcontainer efficiency

### PR DESCRIPTION
**Pull request by richardstartin**
- change `RunContainer.intersects(BitmapContainer)` to check if the bitmap intersects with the range
- change `RunContainer.andCardinality(BitmapContainer)` to add the cardinality in the range of the run
- some miscellaneous test coverage enhancements

### CODE REVIEW

**Comment by Cheappie** [comment]
Hi, @richardstartin I have a few questions to get some context. What is the reason that
this method 'boolean intersects(int minimum, int supremum)' cannot be reused in place of 'boolean isNonEmptyInRange(int start, int end)' in BitmapContainer ?

I see that those methods differently calculate last word, are both correct ?

boolean intersects(int minimum, int supremum): 
'int end = supremum >>> 6;'
boolean isNonEmptyInRange(int start, int end): 
'int endword = (end - 1) >>> 6;'

**Comment by richardstartin** [comment]
Yes good point, thanks, this is a WIP PR.

**Comment by richardstartin** [comment]
@lemire what do you want to do about this LGTM failure? It fails because

```
[2020-04-20 10:20:28] [autobuild] > Could not resolve all files for configuration ':jmh:jmh'.
[2020-04-20 10:20:28] [autobuild]    > Could not find it.uniroma3.mat:extendedset:1.3.4.
[2020-04-20 10:20:28] [autobuild]      Searched in the following locations:
[2020-04-20 10:20:28] [autobuild]        - https://jcenter.bintray.com/it/uniroma3/mat/extendedset/1.3.4/extendedset-1.3.4.pom
[2020-04-20 10:20:28] [autobuild]        - https://metamx.artifactoryonline.com/metamx/pub-libs-releases-local/it/uniroma3/mat/extendedset/1.3.4/extendedset-1.3.4.pom
[2020-04-20 10:20:28] [autobuild]      Required by:
[2020-04-20 10:20:28] [autobuild]          project :jmh
```

Should this be [com.n3twork.druid:extendedset:1.3.4](https://mvnrepository.com/artifact/com.n3twork.druid/extendedset/1.3.4) instead?

**Comment by lemire** [comment]
I am asking for a rerun. It looks like the tests failed.

**Comment by richardstartin** [comment]
@lemire the tests pass, but LGTM breaks looking for dependencies.

**Comment by Cheappie** [comment]
In build logs we can find that both urls which should point to extendedset pom, returns 404 not found error. I would suggest to try changing url in build.gradle.kts file:

instead of
line 12: url = URI("https://metamx.artifactoryonline.com/metamx/pub-libs-releases-local")

try use this \/
url = URI("https://metamx.artifactoryonline.com/artifactory/pub-libs-releases-local")

**Comment by lemire** [comment]
@Cheappie  
I will try to change the URI name on top of Richard's work. Should be harmless enough.

**Comment by lemire** [comment]
Oh. It is a bit complicated because it is a foreign PR.

File: RoaringBitmap/src/main/java/org/roaringbitmap/RunContainer.java [hunk 1]
```diff
@@ -492,14 +492,9 @@ public int andCardinality(BitmapContainer x) {
...
-      int runEnd = runStart + (this.getLength(rlepos));
-      for (int runValue = runStart; runValue <= runEnd; ++runValue) {
-        if (x.contains((char) runValue)) {// it looks like contains() should be cheap enough if
-                                           // accessed sequentially
-          cardinality++;
-        }
-      }
+      int runStart = this.getValue(rlepos);
+      int runEnd = runStart + this.getLength(rlepos);
+      cardinality += x.cardinalityInRange(runStart, runEnd + 1);
```
Comment (lemire) [inline RoaringBitmap/src/main/java/org/roaringbitmap/RunContainer.java:497]: good.

File: RoaringBitmap/src/main/java/org/roaringbitmap/RunContainer.java [hunk 2]
```diff
@@ -745,8 +740,8 @@ public boolean contains(char x) {
     }
     index = -index - 2; // points to preceding value, possibly -1
     if (index != -1) {// possible match
-      int offset = (x) - (getValue(index));
-      int le = (getLength(index));
+      int offset = x - getValue(index);
```
Comment (lemire) [inline RoaringBitmap/src/main/java/org/roaringbitmap/RunContainer.java:743]: good cleaning.

File: jmh/build.gradle.kts [hunk 3]
```diff
@@ -24,7 +24,7 @@ dependencies {
             project(":RoaringBitmap"),
             "com.google.guava:guava:${deps["guava"]}",
             "com.googlecode.javaewah:JavaEWAH:1.0.8",
-            "it.uniroma3.mat:extendedset:1.3.4",
+            "com.n3twork.druid:extendedset:1.3.4",
```
Comment (lemire) [inline jmh/build.gradle.kts:27]: What is the reason for this change? If we want to "upgrade", we could go with org.apache.druid:extendedset:0.18.0 which is far more recent and has been actively maintained.

**Comment by lemire** [comment]
@Cheappie 

> please revert package to 'it.uniroma3.mat'

Why should he revert? I don't understand why he made the change, but I don't care too much that he reverts it back. As far as I can tell, this is an old package that has not received much love over the years... while others have been updating it... so I feel totally warranted to update the dependency if it is somehow useful... I just think we should know why @richardstartin wants to do so.

**Comment by Cheappie** [comment]
@lemire I thought we aim to repair finding that dependency, but if there is any better alternative then ofc let's use it

**Comment by lemire** [comment]
Ok. Let us wait for @richardstartin  to ping us back.

**Comment by Cheappie** [comment]
@lemire @richardstartin 
Would it be possible to make new release to maven central, after this change is merged ? I am asking because it significantly improves performance of zero-gc cardinality calculations between run and bitmap containers

File: jmh/build.gradle.kts (hunk 3 above)
Comment (richardstartin) [reply jmh/build.gradle.kts:27]: resolved

**Comment by richardstartin** [comment]
@lemire there's a lot more low hanging fruit for run vs bitmap. Do you want to merge now or batch in some more improvements?

**Comment by richardstartin** [comment]
Let's fix LGTM another time (I couldn't care less).

**Comment by lemire** [comment]
This has been merged and released.

**Comment by lemire** [comment]
@richardstartin I missed your comment about low-hanging fruits. Can you elaborate?
